File: androidApp/src/main/java/com/agrichain/weatherapp/android/WeatherScreen.kt
package com.agrichain.weatherapp.android

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.agrichain.weatherapp.forecast.Forecast
import com.agrichain.weatherapp.forecast.ForecastViewModel
import org.koin.androidx.compose.koinViewModel

@Composable
fun WeatherScreen(
    forecastViewModel: ForecastViewModel = koinViewModel()
) {

    val forecastState = forecastViewModel.forecastState.collectAsState()

    if (forecastState.value.loading) {
        Box(
            modifier = Modifier.fillMaxSize(),
            contentAlignment = Alignment.Center
        ) {
            CircularProgressIndicator()
        }
    }

    if (forecastState.value.error != null) {
        Box(
            modifier = Modifier.fillMaxSize(),
            contentAlignment = Alignment.Center
        ) {
            Text(
                text = forecastState.value.error!!
            )
        }
    }

    if (forecastState.value.forecast != null) {
        ForecastScreen(
            forecast = forecastState.value.forecast!!,
            onRefresh = {
                forecastViewModel.getForecast(true)
            }
        )
    }


}

@Composable
fun ForecastScreen(
    forecast: Forecast,
    onRefresh: () -> Unit
) {

    Scaffold(
        modifier = Modifier.fillMaxSize(),
        content = { padding ->
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(padding)
                    .padding(16.dp),
                contentAlignment = Alignment.Center
            ) {
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp),
                    shape = RoundedCornerShape(16.dp),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.surface
                    ),
                    elevation = CardDefaults.cardElevation(defaultElevation = 4.dp)
                ) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(24.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(16.dp)
                    ) {

                        Button(
                            onClick = {
                                onRefresh()
                            },
                        ) {
                            Text(
                                text = "Refresh"
                            )
                        }

                        // Location and Time
                        Text(
                            text = forecast.timezone,
                            style = MaterialTheme.typography.headlineMedium,
                            fontWeight = FontWeight.Bold
                        )
                        Text(
                            text = "${forecast.latitude}, ${forecast.longitude} | ${forecast.time}",
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )

                        // Weather Icon and Temperature
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(8.dp)
                        ) {
                            Text(
                                text = forecast.weatherIcon,
                                fontSize = 48.sp,
                                modifier = Modifier.padding(end = 8.dp)
                            )
                            Column {
                                Text(
                                    text = forecast.temperature,
                                    style = MaterialTheme.typography.displayMedium,
                                    fontWeight = FontWeight.Bold
                                )
                                Text(
                                    text = forecast.weather,
                                    style = MaterialTheme.typography.titleMedium,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant
                                )
                            }
                        }

                        // Additional Weather Details
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceEvenly
                        ) {
                            WeatherDetailItem(
                                label = "Wind Speed",
                                value = forecast.windSpeed
                            )
                            WeatherDetailItem(
                                label = "Wind Direction",
                                value = forecast.windDirection
                            )
                            WeatherDetailItem(
                                label = "Elevation",
                                value = "${forecast.elevation} m"
                            )
                        }

                        // Timezone and Day/Night
                        Text(
                            text = "Timezone: ${forecast.timezoneAbbreviation} (${if (forecast.isDay.toBoolean()) "Day" else "Night"})",
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            textAlign = TextAlign.Center
                        )
                    }
                }
            }
        }
    )
}

@Composable
fun WeatherDetailItem(label: String, value: String) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
        Text(
            text = value,
            style = MaterialTheme.typography.bodyLarge,
            fontWeight = FontWeight.Medium
        )
    }
}

File: iosApp/iosApp/WeatherScreen.swift
import SwiftUI
import shared

extension WeatherScreen {
    
    @MainActor
    class ForecastViewModelWrapper: ObservableObject {
        let forecastViewModel: ForecastViewModel
        
        init() {
            forecastViewModel = ForecastInjector().forecastViewModel
            forecastState = forecastViewModel.forecastState.value
        }
        
        @Published var forecastState: ForecastState
        
        func startObserving() {
            Task {
                for await stateF in forecastViewModel.forecastState {
                    self.forecastState = stateF
                }
            }
        }
    }
}

struct WeatherScreen: View {
    
    @ObservedObject private(set) var viewModel: ForecastViewModelWrapper
    
    var body: some View {
        
        VStack {
            if viewModel.forecastState.loading {
                Loader()
            }
            
            if let error = viewModel.forecastState.error {
                ErrorMessage(message: error)
            }
            
            if let forecast = viewModel.forecastState.forecast {
                ForecastScreen(forecast: forecast, onRefresh: { viewModel.forecastViewModel.getForecast(forceRefresh: true) })
            }
        }.onAppear{
            self.viewModel.startObserving()
        }
        
    }
}

struct ForecastScreen: View {
    let forecast: Forecast
    let onRefresh: () -> Void
    
    var body: some View {
        ZStack {
            Color(.systemBackground)
                .ignoresSafeArea()
            
            VStack {
                
                Button(action: {
                    onRefresh()
                }) {
                    Text("Refresh")
                }
                
                // Card container
                ZStack {
                    RoundedRectangle(cornerRadius: 16)
                        .fill(Color(.systemBackground))
                        .shadow(radius: 4)
                    
                    VStack(spacing: 16) {
                        // Location and Time
                        Text(forecast.timezone)
                            .font(.title)
                            .fontWeight(.bold)
                        Text("\(forecast.latitude), \(forecast.longitude) | \(forecast.time)")
                            .font(.subheadline)
                            .foregroundColor(.secondary)
                        
                        // Weather Icon and Temperature
                        HStack(spacing: 8) {
                            Text(forecast.weatherIcon)
                                .font(.system(size: 48))
                            VStack(alignment: .leading) {
                                Text(forecast.temperature)
                                    .font(.system(size: 48, weight: .bold))
                                Text(forecast.weather)
                                    .font(.title3)
                                    .foregroundColor(.secondary)
                            }
                        }
                        .padding(.top, 8)
                        
                        // Additional Weather Details
                        HStack(spacing: 16) {
                            WeatherDetailItem(
                                label: "Wind Speed",
                                value: forecast.windSpeed
                            )
                            WeatherDetailItem(
                                label: "Wind Direction",
                                value: forecast.windDirection
                            )
                            WeatherDetailItem(
                                label: "Elevation",
                                value: "\(forecast.elevation) m"
                            )
                        }
                        .padding(.vertical, 8)
                        
                        // Timezone and Day/Night
                        Text("Timezone: \(forecast.timezoneAbbreviation) (\((forecast.isDay == "1") ? "Day" : "Night"))")
                            .font(.caption)
                            .foregroundColor(.secondary)
                    }
                    .padding(24)
                }
                .padding(.horizontal, 16)
                
                Spacer()
            }
            .padding(.top, 16)
        }.refreshable {
            onRefresh() // Trigger the refresh action
        }
    }
}

struct WeatherDetailItem: View {
    let label: String
    let value: String
    
    var body: some View {
        VStack {
            Text(label)
                .font(.caption)
                .foregroundColor(.secondary)
            Text(value)
                .font(.body)
                .fontWeight(.medium)
        }
    }
}

struct Loader: View {
    var body: some View {
        ProgressView()
    }
}

struct ErrorMessage: View {
    var message: String
    
    var body: some View {
        Text(message)
            .font(.title)
    }
}
